pkg/knownhosts/knownhosts.go

```diff
@@ -0,0 +1,350 @@
+)
+
+type KnownHosts struct {
+	Lines []*Line
+}
+
+func Unmarshal(in io.Reader) (*KnownHosts, error) {
+	k := &KnownHosts{
+		Lines: []*Line{},
+	}
+	s := bufio.NewScanner(in)
+	for s.Scan() {
+		l, err := parseHosts(s.Text())
+		if err == nil && l != nil {
+			k.Lines = append(k.Lines, l)
+		} else {
+			fmt.Println(err)
+		}
+	}
+	return k, s.Err()
+}
+
+func (k *KnownHosts) Marshal(out io.Writer) error {
+	for _, l := range k.Lines {
+		b, err := l.Marshal()
+		if err != nil {
+			return err
+		}
+		if _, err := out.Write(b); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+var HostNotFoundError = errors.New("Host is unknown")
+var HostRevokedError = errors.New("Host is revoked")
+var UnsupportedAddrType = errors.New("remote must be a *net.TCPAddr")
+
+func (k *KnownHosts) HostKeyCallback(hostname string, remote net.Addr, key ssh.PublicKey) error {
+	var addr *net.TCPAddr
+	if v, ok := remote.(*net.TCPAddr); ok {
+		addr = v
+	} else {
+		return UnsupportedAddrType
+	}
+	keyBytes := key.Marshal()
+	var matched = []*Host{}
+	for _, l := range k.Lines {
+		if l.CertAuthority {
+			continue
+		}
+		for _, h := range l.Hosts {
+			if h.Match(hostname, addr) && bytes.Equal(keyBytes, l.PublicKey.Marshal()) {
+				if l.Revoked {
+					return HostRevokedError
+				}
+				matched = append(matched, h)
+			}
+		}
+	}
+	if len(matched) == 0 {
+		return HostNotFoundError
+	}
+	return nil
+}
+
+func (k *KnownHosts) AddHost(hostname string, key ssh.PublicKey) error {
+	_parts := strings.Split(hostname, ":")
+	hostname = _parts[0]
+	port := 22
+	if len(_parts) > 1 && _parts[1] != "" {
+		var err error
+		port, err = strconv.Atoi(_parts[1])
+		if err != nil {
+			return err
+		}
+	}
+	l := &Line{
+		PublicKey: key,
+	}
+	l.Hosts = []*Host{{
+		Addr: hostname,
+		Port: port,
+		line: l,
+	}}
+	k.Lines = append(k.Lines, l)
+	return nil
+}
+
+type Line struct {
+	Hosts         []*Host
+	Revoked       bool
+	CertAuthority bool
+	PublicKey     ssh.PublicKey
+}
+
+func (l *Line) Marshal() ([]byte, error) {
+	const sep = " "
+	var buf bytes.Buffer
+
+	// flag part
+	var flag string
+	if l.Revoked {
+		flag = "revoked"
+	}
+	if l.CertAuthority {
+		flag = "cert-authority"
+	}
+	if flag != "" {
+		buf.WriteString(fmt.Sprintf("@%s", flag))
+		buf.WriteString(sep)
+	}
+
+	// hosts part
+	for i, h := range l.Hosts {
+		if h.Hash != nil && (i != 0 || len(l.Hosts) > 1) {
+			return nil, fmt.Errorf("Hashed host must be only host on line, found %d!", len(l.Hosts))
+		}
+		if b := h.Marshal(); b != nil {
+			buf.Write(b)
+		} else {
+			return nil, fmt.Errorf("Invalid host: %#v", h)
+		}
+		if i != len(l.Hosts)-1 {
+			buf.WriteString(",")
+		}
+	}
+	buf.WriteString(sep)
+
+	// key part (has trailing newline)
+	buf.Write(ssh.MarshalAuthorizedKey(l.PublicKey))
+
+	return buf.Bytes(), nil
+}
+
+type Host struct {
+	Hash           *hostHash
+	Pattern        *regexp.Regexp
+	PatternNegated bool
+	PatternSource  string
+	Addr           string
+	Port           int
```

Comment:
Use string for ports, everything that works with them uses them as a string, so the round-trip doesn't help anything.